### PR TITLE
gh-103165: prevent moving of folders into child folders in case insen…

### PR DESCRIPTION
**Pull request by Jokimax**
…tive operating systems

* Issue: gh-103165

### CODE REVIEW

File: Lib/shutil.py [hunk 1]
```diff
@@ -865,7 +865,10 @@ def _destinsrc(src, dst):
         src += os.path.sep
     if not dst.endswith(os.path.sep):
         dst += os.path.sep
-    return dst.startswith(src)
+    try:
+        return os.lstat(src) == os.lstat(dst[:len(src)])
```
Comment (serhiy-storchaka) [inline Lib/shutil.py]: It is better to use `os.path.samestat()`.

File: Lib/shutil.py [hunk 2]
```diff
@@ -865,7 +865,10 @@ def _destinsrc(src, dst):
...
+    except OSError:
+        return dst.startswith(src)
```
Comment (serhiy-storchaka) [inline Lib/shutil.py]: In what cases this fallback is needed? And how can it be used if the original issue is that it doesn't work on case-insensitive FS?

Could you please add a test? This is a non-trivial issue and it can have many corner cases.
